File: bootstrap/src/main/java/voodoo/Bootstrap.java
package voodoo;

import org.w3c.dom.Document;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.xpath.XPath;
import javax.xml.xpath.XPathConstants;
import javax.xml.xpath.XPathFactory;
import java.io.*;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.nio.file.StandardCopyOption;
import java.security.MessageDigest;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Properties;
import java.util.stream.Collectors;

public class Bootstrap {
    private static final File binariesDir = new File("../.voodoo/");
    private static final File lastFile = new File(binariesDir, "newest.jar");
    private static final Properties props = new Properties();

    static {
        try {
            props.load(Bootstrap.class.getResourceAsStream("/maven.properties"));
        } catch (IOException e) {
            e.printStackTrace();
            System.exit(-1);
        }
    }

    private static final String mavenUrl = props.getProperty("url");
    private static final String group = props.getProperty("group");
    private static final String artifact = props.getProperty("name");
    private static final String classifier = props.getProperty("classifier");

    public static void main(String[] args) {
        try {
            cleanup();
            launch(args);
        } catch(Throwable t) {
            t.printStackTrace();
            System.err.println("Error: " + t.getLocalizedMessage());
            System.exit(-1);
        }
    }

    private static void cleanup() {
        binariesDir.mkdirs();
        File[] files = binariesDir.listFiles((pathname -> pathname.getName().endsWith(".tmp")));
        for (File file : files) {
            file.delete();
        }
    }


    private static void launch(String[] originalArgs) throws Throwable {
        System.out.printf("Downloading the %s binary...%n", artifact);

        File file;
        try {
            file = download();
            if(!file.exists()) {
                throw new IllegalStateException("downloaded file does not seem to exist");
            }
        } catch(IOException e) {
            e.printStackTrace();
            System.err.printf("cannot download %s from %s, trying to reuse last binary%n", artifact, mavenUrl);
            file = lastFile;
        }

        if(!file.exists()) {
            throw new IllegalStateException(String.format("binary %s does not exist", file.getPath()));
        }
        System.out.printf("Loaded %s%n", file.getPath());
        String java = Paths.get(System.getProperty("java.home"), "bin", "java").toFile().getPath();
        File workingDir = new File(System.getProperty("user.dir"));

        String[] debugArgs;
        if(System.getProperty("kotlinx.coroutines.debug") != null) {
            debugArgs = new String[]{"-Dkotlinx.coroutines.debug"};
        } else {
            debugArgs = new String[0];
        }

        ArrayList<String> argsList = new ArrayList<>();
        argsList.add(java);
        argsList.addAll(Arrays.asList(debugArgs));
        argsList.add("-jar");
        argsList.add(file.getPath());
        argsList.addAll(Arrays.asList(originalArgs));

        String[] args = argsList.toArray(new String[0]);

        System.out.printf("Executing %s", argsList.toString());
        int exitStatus = new ProcessBuilder(args)
            .directory(workingDir)
                .redirectOutput(ProcessBuilder.Redirect.INHERIT)
                .redirectError(ProcessBuilder.Redirect.INHERIT)
                .start()
                .waitFor();
        System.exit(exitStatus);
    }

    private static File download() throws Exception {
        String groupPath = group.replace( '.', '/');
        String mavenMetadataUrl = mavenUrl + '/' + groupPath + '/' + artifact + "/maven-metadata.xml";
        DocumentBuilderFactory dbFactory = DocumentBuilderFactory.newInstance();
        DocumentBuilder dBuilder = dbFactory.newDocumentBuilder();
        Document doc = dBuilder.parse((new URL(mavenMetadataUrl)).openStream());
        XPathFactory xpFactory = XPathFactory.newInstance();
        XPath xPath = xpFactory.newXPath();
        String xpath = "/metadata/versioning/release/text()";
        String releaseVersion = (String) xPath.evaluate(xpath, doc, XPathConstants.STRING);
        return downloadArtifact(mavenUrl, group, artifact, releaseVersion, classifier, "jar", binariesDir);
    }

    public static byte[] createMD5(File file) throws Exception {
        InputStream fis =  new FileInputStream(file);

        byte[] buffer = new byte[1024];
        MessageDigest complete = MessageDigest.getInstance("MD5");
        int numRead;

        do {
            numRead = fis.read(buffer);
            if (numRead > 0) {
                complete.update(buffer, 0, numRead);
            }
        } while (numRead != -1);

        fis.close();
        return complete.digest();
    }

    private static final String toHexString(byte[] bytes) {
        StringBuilder sb = new StringBuilder();
        for (byte b : bytes) {
            sb.append(String.format("%02x", b));
        }
        return sb.toString();
    }

    public static  File downloadArtifact(String mavenUrl,
                                         String group,
                                         String artifactId,
                                         String version,
                                         String classifier,
                                         String extension,
                                         File outputDir
    ) throws Exception {
        String groupPath;
        String classifierSuffix;
        groupPath = group.replace('.', '/');
        if (classifier != null) {
            classifierSuffix = '-' + classifier;
        } else {
            classifierSuffix = "";
        }

        String artifactUrl = mavenUrl + '/' + groupPath + '/' + artifactId + '/' + version + '/' + artifactId + '-' + version + classifierSuffix + '.' + extension;
        File tmpFile = new File(outputDir, artifactId + '-' + version + classifierSuffix + '.' + extension + ".tmp");
        File targetFile = new File(outputDir, artifactId + '-' + version + classifierSuffix + '.' + extension);

        String md5;
        try (BufferedReader reader = new BufferedReader(new InputStreamReader(new URL(artifactUrl + ".md5").openStream(), StandardCharsets.UTF_8))) {
            md5 = reader.lines().collect(Collectors.joining(System.lineSeparator()));
        }

        if (targetFile.exists()) {
            String fileMd5 = toHexString(createMD5(targetFile));

            if (fileMd5.equalsIgnoreCase(md5)) {
                System.out.println("cached file matched md5 hash");
                return targetFile;
            }
        }

        try (BufferedInputStream in = new BufferedInputStream(new URL(artifactUrl).openStream());
             FileOutputStream fileOutputStream = new FileOutputStream(tmpFile)) {
            byte[] dataBuffer = new byte[1024];
            int bytesRead;
            while ((bytesRead = in.read(dataBuffer, 0, 1024)) != -1) {
                fileOutputStream.write(dataBuffer, 0, bytesRead);
            }
        }

        {
            String fileMd5 = toHexString(createMD5(tmpFile));
            if(!fileMd5.equalsIgnoreCase(md5)) {
                throw new IllegalArgumentException(String.format("%s did not match md5 hash: '%s' file: %s", artifactUrl, md5, fileMd5));
            }
        }

        Files.copy(tmpFile.toPath(), targetFile.toPath(), StandardCopyOption.REPLACE_EXISTING);
        tmpFile.delete();

        return targetFile;
    }

}

File: bootstrap/src/main/kotlin/voodoo/Bootstrap.kt
///*
// * SK's Minecraft Launcher
// * Copyright (C) 2010-2014 Albert Pham <http://www.sk89q.com> and contributors
// * Please see LICENSE.txt for license information.
// */
//
//package voodoo
//
//import java.io.File
//import java.io.FileOutputStream
//import java.io.IOException
//import java.net.URL
//import java.security.MessageDigest
//import java.util.*
//import java.util.stream.Collectors
//import javax.xml.parsers.DocumentBuilderFactory
//import javax.xml.xpath.XPathConstants
//import javax.xml.xpath.XPathFactory
//import kotlin.system.exitProcess
//
//
//@Throws(Throwable::class)
//fun main(vararg args: String) {
//    try {
//        Bootstrap.cleanup()
//        Bootstrap.launch(*args)
//    } catch (t: Throwable) {
//        t.printStackTrace()
//        System.err.println("Error")
//        exitProcess(-1)
//    }
//}
//
//object Bootstrap {
//
//    val props = Properties().apply {
//        javaClass.getResourceAsStream("/maven.properties").bufferedReader().use {
//            load(it)
//        }
//    }
//
//    private val mavenUrl = props.getProperty("url") as String
//    private val group = props.getProperty("group") as String
//    private val artifact = props.getProperty("name") as String
//    private val classifier = props.getProperty("classifier") as String
//
////    private val directories: Directories = Directories.get(moduleName = "$artifact-bootstrap")
//    private val binariesDir: File = File(".voodoo") // directories.cacheHome
//    private val lastFile: File = binariesDir.resolve("newest.jar")
//
//    fun cleanup() {
//        val files = binariesDir.listFiles { pathname -> pathname.name.endsWith(".tmp") }
//
//        if (files != null) {
//            for (file in files) {
//                file.delete()
//            }
//        }
//    }
//
//    private fun download(): File {
//
//        val groupPath = group.replace('.','/')
//        val mavenMetadataUrl = "$mavenUrl/$groupPath/$artifact/maven-metadata.xml"
//
//        val dbFactory = DocumentBuilderFactory.newInstance()
//        val dBuilder = dbFactory.newDocumentBuilder()
////        val xmlInput = InputSource(StringReader(xmlText))
//        val doc = dBuilder.parse(URL(mavenMetadataUrl).openStream())
//        val xpFactory = XPathFactory.newInstance()
//        val xPath = xpFactory.newXPath()
//
//
//        val xpath = "/metadata/versioning/release/text()"
//        val releaseVersion = xPath.evaluate(xpath, doc, XPathConstants.STRING) as String
//
//        // TODO: select latest build within major or minor version ?
//
//        val artifactFile = downloadArtifact(
//            mavenUrl = mavenUrl,
//            group = group,
//            artifactId = artifact,
//            version = releaseVersion,
//            classifier = classifier,
//            extension = "jar",
//            outputDir = binariesDir
//        )
//
//        return artifactFile
//    }
//
//    @Throws(Throwable::class)
//    fun launch(vararg originalArgs: String) {
//        println("Downloading the $artifact binary...")
//        val file = try {
//            download().apply {
//                require(exists()) { "downloaded files does not seem to exist" }
//                copyTo(lastFile, overwrite = true)
//            }
//        } catch (e: IOException) {
//            e.printStackTrace()
//            System.err.println("cannot download $artifact from $mavenUrl, trying to reuse last binary")
//            lastFile
//        }
//
//        require(file.exists()) { "binary $file does not exist" }
//
//        println("Loaded " + file.path)
//        val java = arrayOf(System.getProperty("java.home"), "bin", "java").joinToString(File.separator)
//        val workingDir = File(System.getProperty("user.dir"))
//
//        val debugArgs = System.getProperty("kotlinx.coroutines.debug")?.let {
//            arrayOf("-Dkotlinx.coroutines.debug")
//        } ?: emptyArray()
//
//        val args = arrayOf(java, *debugArgs, "-jar", file.path, *originalArgs)
//        println("executing [${args.joinToString { "'$it'" }}]")
//        val exitStatus = ProcessBuilder(*args)
//            .directory(workingDir)
//            .redirectOutput(ProcessBuilder.Redirect.INHERIT)
//            .redirectError(ProcessBuilder.Redirect.INHERIT)
//            .start()
//            .waitFor()
//        exitProcess(exitStatus)
//    }
//
//    fun downloadArtifact(
//        mavenUrl: String,
//        group: String,
//        artifactId: String,
//        version: String,
//        classifier: String? = null,
//        extension: String = "jar",
//        outputDir: File,
//        outputFile: File? = null
//    ): File {
//        val groupPath = group.replace('.', '/')
//
//        val classifierSuffix = classifier?.let { "-$it"} ?: ""
//        val artifactUrl = "$mavenUrl/$groupPath/$artifactId/$version/$artifactId-$version$classifierSuffix.$extension"
//        val tmpFile = File(outputDir, "$artifactId-$version$classifierSuffix.$extension.tmp")
//        val targetFile = outputFile ?: File(outputDir, "$artifactId-$version$classifierSuffix.$extension")
//
//        val md5 = URL("$artifactUrl.md5").openStream().bufferedReader().use { bis ->
//            bis.lines().collect(Collectors.joining(System.lineSeparator()))
//        }
//
//        if(targetFile.exists()) {
//            val md = MessageDigest.getInstance("MD5")
//            md.update(targetFile.readBytes())
//            val digest = md.digest()
//            val fileMd5 = digest.toHexString() // DigestUtils.md5Hex(tmpFile.inputStream())
//            if(fileMd5.toLowerCase() == md5.toLowerCase()) {
//                println("cached file matched md5 hash")
//                return targetFile
//            }
//        }
//
//        run {
//            val url = URL(artifactUrl)
//            url.openStream().buffered().use { bis ->
//                FileOutputStream(tmpFile).use { fis ->
//                    val buffer = ByteArray(1024)
//                    var count = 0
//                    while (bis.read(buffer, 0, 1024).also { count = it } != -1) {
//                        fis.write(buffer, 0, count)
//                    }
//                }
//            }
//        }
//
//
//        run {
//            val md = MessageDigest.getInstance("MD5")
//            md.update(tmpFile.readBytes())
//            val digest = md.digest()
//            val fileMd5 = digest.toHexString() // DigestUtils.md5Hex(tmpFile.inputStream())
//            require(fileMd5.toLowerCase() == md5.toLowerCase()) { "$artifactUrl did not match md5 hash: '$md5' file: $fileMd5" }
//        }
//
//        tmpFile.copyTo(targetFile, overwrite = true)
//        tmpFile.delete()
//        return targetFile
//    }
//    fun ByteArray.toHexString(): String = joinToString("", transform = { "%02x".format(it) })
//}
